LSPS2/README.md

```diff
@@ -0,0 +1,736 @@
+  step:
+  * The current time is strictly less than
+    `opening_fee_params.valid_until`.
+  * All received payment parts sum up to at least `payment_size`
+    plus `opening_fee`.
+  * The client is connected.
+
+In "no-MPP+var-invoice" mode, the LSP, if it receives a forward
+where the next hop is `jit_channel_scid`, before
+`opening_fee_params.valid_until`:
+
+* MUST set `payment_size` as the forwarded value of the HTLC.
+* MUST compute the `opening_fee` based on the `payment_size`, and
+  if the computation overflows, MUST fail with `unknown_next_peer`.
+* MUST check that `opening_fee + htlc_minimum_msat < payment_size`,
+  and if that fails, MUST fail with `unknown_next_peer`.
+
+### 5.  Channel Opening And Forwarding
+
+The LSP requests a channel open to the client.
+
+The LSP selects the channel size.
+The LSP SHOULD scale the size of the channel based on the
+`payment_size`.
+The LSP MUST ensure the channel is large enough to transmit the
+`payment_size - opening_fee`, plus any reserve needed on the LSP
+side.
+
+The LSP and client MUST negotiate these options and parameters:
+
+* `option_zeroconf` is set.
+* `option_scid_alias` is set.
+* `announce_channel` is `false`.
+* The client-provided `to_self_delay` is less than or equal to
+  the agreed-upon `max_client_to_self_delay`.
+
+> **Rationale** `option_zeroconf` is set so that the payment can
+> be immediately forwarded to the client and resolved as soon as
+> the opening negotiation completes; if the client and LSP had to
+> wait for confirmation, the incoming payment would have to be
+> locked for an extended period, which is indistinguishable from
+> a channel jamming attack.
+> `option_zeroconf` cannot be set unless `announce_channel` is
+> `false`.
+> `option_scid_alias` needs to be set so that the channel can be
+> referred to on future invoices before the channel confirms.
+>
+> Future versions of this API may be able to utilize Asynchronous
+> Receive (currently being developed as of this version) to
+> be able to get around the `option_zeroconf` requirement, by
+> treating the receiver / client as offline until the client is
+> online *and* the JIT channel to it is deeply confirmed, if the
+> LSP and client negotiate a non-`option_zeroconf` channel for
+> the JIT channel.
+> A non-`option_zeroconf` JIT channel could also be later
+> published with `announce_channel = true`.
+
+Clients MAY reject the channel open attempt if any other
+parameters or options are not desired by the client (for example,
+it MAY reject non-zero channel reserve requirements imposed on
+the client by the LSP).
+Rejection is via the `error` Lightning BOLT message.
+
+LSPs MUST fail the payment with `unknown_next_peer` in case the
+client rejects the channel open via an `error`.
+LSPs MUST NOT treat client disconnection during the opening
+protocol as a rejection, but MUST retry if the client
+reconnects before the `valid_until` datetime.
```

Comment:
Not immediately I think; it can simply wait for the client to come online again.  In particular `unknown_next_peer` has the `PERM` bit set, which would cause the payer to give up on this SCID and assume it is permanently broken/closed/notexist, so giving that error would stop future payment attempts.

Maybe `temporary_channel_error` if you really want to fail the payment here.

Reply:
Oh wow, I didn't know unknown_next_peer was PERM. That's returned by lightning nodes whenever the peer is offline  (temporarily) right? 

I think waiting to come online is not a good idea unless there is some notification system involved. We don't want to be locking up liquidity unnecessarily.

Probably the error code doesn't even really matter much. Both with `unknown_next_peer` and `temporary_channel_failure` the sender would ignore this channel for a too long period. I'm not sure how exactly the different implementations handle this. If they don't remember the channel in the hop hint was unavailable, unknown_next_peer would work too.

Reply:
>That's returned by lightning nodes whenever the peer is offline (temporarily) right?

It should be `temporary_channel_failure` I think; after all, you *do* know the peer, it is just not available right now (which might be a temporary thing).

> I'm not sure how exactly the different implementations handle this. If they don't remember the channel in the hop hint was unavailable, unknown_next_peer would work too.

`PERM` errors would cause some versions of CLN to actively delete the failing channel / node from the gossip map.  But then CLN does not actually put routehints into the gossip map (it routefinds to the entry point of the routehint, then tacks on the routehint at the end of the route).

But if we go with what BOLT says, we should probably use `temporary_channel_failure` for this.

Reply:
Clarified to raise `temporary_channel_failure` instead if the client disconnects during channel opening.

Reply:
Interesting, so you think LND handles this the wrong way? That returns unknown next peer any time a peer is offline right?

Reply:
`unknown_next_peer` is `PERM`, which implies you should remove it from the routemap.  The intended use is if the last channel with that peer has been closed (`shutdown` has been sent, which starts the mutual closing process, which can take a long time actually as both sides negotiate a single closing fee and each side can be programmed to hardass and only move the fee by 1 sat at a time) but this fact is not reflected onchain yet (because mutual close notification is being negotiated and there is no mutual close transaction yet).

`unknown_next_peer` for disconnected peers is wrong to me.

Reply:
Seems ok, resolving.